File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/repository/SleepRepository.kt
package com.bissbilanz.repository

import app.cash.sqldelight.coroutines.asFlow
import app.cash.sqldelight.coroutines.mapToList
import com.bissbilanz.ErrorReporter
import com.bissbilanz.api.BissbilanzApi
import com.bissbilanz.api.generated.model.SleepCreate
import com.bissbilanz.api.generated.model.SleepEntry
import com.bissbilanz.api.generated.model.SleepFoodCorrelationEntry
import com.bissbilanz.api.generated.model.SleepUpdate
import com.bissbilanz.cache.BissbilanzDatabase
import com.bissbilanz.mode.AppModeManager
import com.bissbilanz.sync.SyncOperation
import com.bissbilanz.sync.SyncQueue
import com.bissbilanz.userdata.UserDataDatabase
import com.bissbilanz.util.decodeOrNull
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.datetime.Clock
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

class SleepRepository(
    private val api: BissbilanzApi,
    private val db: UserDataDatabase,
    private val cacheDb: BissbilanzDatabase,
    private val syncQueue: SyncQueue,
    private val json: Json,
    private val errorReporter: ErrorReporter,
    private val appModeManager: AppModeManager,
) {
    fun entries(): Flow<List<SleepEntry>> =
        db.userDataDatabaseQueries
            .selectAllSleepEntries()
            .asFlow()
            .mapToList(Dispatchers.IO)
            .map { rows -> rows.mapNotNull { json.decodeOrNull<SleepEntry>(it.jsonData) } }

    suspend fun refresh(
        from: String? = null,
        to: String? = null,
    ) {
        if (appModeManager.isLocal) return
        try {
            val entries = api.getSleepEntries(from, to)
            cacheSleepEntries(entries)
        } catch (e: Exception) {
            if (e is kotlinx.coroutines.CancellationException) throw e
            errorReporter.captureException(e)
        }
    }

    suspend fun createEntry(entry: SleepCreate): SleepEntry {
        val temp = sleepCreateToEntry(entry)
        cacheSleepEntry(temp)
        syncQueue.enqueue(SyncOperation.CreateSleep(json.encodeToString(entry), localId = temp.id))
        return temp
    }

    suspend fun updateEntry(
        id: String,
        entry: SleepUpdate,
    ): SleepEntry {
        val cached = db.userDataDatabaseQueries.selectAllSleepEntries().executeAsList()
        val existing = cached.mapNotNull { json.decodeOrNull<SleepEntry>(it.jsonData) }.find { it.id == id }
        val result =
            if (existing != null) {
                val updated =
                    existing.copy(
                        durationMinutes = entry.durationMinutes ?: existing.durationMinutes,
                        quality = entry.quality ?: existing.quality,
                        entryDate = entry.entryDate ?: existing.entryDate,
                        bedtime = entry.bedtime ?: existing.bedtime,
                        wakeTime = entry.wakeTime ?: existing.wakeTime,
                        wakeUps = entry.wakeUps ?: existing.wakeUps,
                        notes = entry.notes ?: existing.notes,
                    )
                cacheSleepEntry(updated)
                updated
            } else {
                SleepEntry(
                    id = id,
                    userId = "",
                    entryDate = entry.entryDate ?: "",
                    durationMinutes = entry.durationMinutes ?: 0,
                    quality = entry.quality ?: 0,
                    bedtime = entry.bedtime,
                    wakeTime = entry.wakeTime,
                    wakeUps = entry.wakeUps,
                    sleepLatencyMinutes = null,
                    deepSleepMinutes = null,
                    lightSleepMinutes = null,
                    remSleepMinutes = null,
                    source = null,
                    notes = entry.notes,
                )
            }
        if (id.startsWith("temp_")) {
            coalesceQueuedCreate(id, entry)
        } else {
            syncQueue.enqueue(SyncOperation.UpdateSleep(id, json.encodeToString(entry)))
        }
        return result
    }

    suspend fun deleteEntry(id: String) {
        db.userDataDatabaseQueries.deleteSleepEntry(id)
        if (id.startsWith("temp_")) {
            syncQueue.removeByAffected("sleep", id)
        } else {
            syncQueue.enqueue(SyncOperation.DeleteSleep(id))
        }
    }

    /**
     * Rewrites the still-queued Create operation for a temp-id sleep entry so the
     * eventual upload carries the edited values. If the create has already been drained
     * (no queued op found), the update is skipped — the temp id is unknown server-side.
     */
    private suspend fun coalesceQueuedCreate(
        tempId: String,
        update: SleepUpdate,
    ) {
        for (req in syncQueue.findByAffected("sleep", tempId)) {
            val create = req.operation as? SyncOperation.CreateSleep ?: continue
            val body = json.decodeOrNull<SleepCreate>(create.body) ?: continue
            val merged =
                body.copy(
                    durationMinutes = update.durationMinutes ?: body.durationMinutes,
                    quality = update.quality ?: body.quality,
                    entryDate = update.entryDate ?: body.entryDate,
                    bedtime = update.bedtime ?: body.bedtime,
                    wakeTime = update.wakeTime ?: body.wakeTime,
                    wakeUps = update.wakeUps ?: body.wakeUps,
                    notes = update.notes ?: body.notes,
                )
            syncQueue.replaceOperation(req.id, create.copy(body = json.encodeToString(merged)))
        }
    }

    /**
     * Returns a single sleep entry by [id]. Checks the local cache first; if not
     * present, performs a range refresh and looks again. Returns null when the
     * entry cannot be found on the server either.
     */
    suspend fun getEntry(id: String): SleepEntry? {
        findInCache(id)?.let { return it }
        refresh()
        return findInCache(id)
    }

    private fun findInCache(id: String): SleepEntry? =
        db.userDataDatabaseQueries
            .selectAllSleepEntries()
            .executeAsList()
            .asSequence()
            .mapNotNull { json.decodeOrNull<SleepEntry>(it.jsonData) }
            .firstOrNull { it.id == id }

    suspend fun getSleepFoodCorrelation(
        startDate: String,
        endDate: String,
    ): List<SleepFoodCorrelationEntry> {
        // Server-side analytics; the UI hides this in Local mode.
        if (appModeManager.isLocal) return emptyList()
        return try {
            api.getSleepFoodCorrelation(startDate, endDate)
        } catch (e: Exception) {
            if (e is kotlinx.coroutines.CancellationException) throw e
            errorReporter.captureException(e)
            emptyList()
        }
    }

    private fun cacheSleepEntry(entry: SleepEntry) {
        db.userDataDatabaseQueries.insertSleepEntry(
            id = entry.id,
            entryDate = entry.entryDate,
            durationMinutes = entry.durationMinutes.toLong(),
            quality = entry.quality.toLong(),
            loggedAt = entry.loggedAt,
            jsonData = json.encodeToString(entry),
        )
    }

    private fun cacheSleepEntries(entries: List<SleepEntry>) {
        db.userDataDatabaseQueries.transaction {
            db.userDataDatabaseQueries.deleteAllSleepEntries()
            entries.forEach { entry -> cacheSleepEntry(entry) }
        }
        // SyncMeta lives in the cache database; written after the user-data commit.
        cacheDb.bissbilanzDatabaseQueries.upsertSyncMeta(
            entityType = "sleep",
            lastSyncedAt = Clock.System.now().toString(),
        )
    }

    @OptIn(ExperimentalUuidApi::class)
    private fun sleepCreateToEntry(entry: SleepCreate): SleepEntry =
        SleepEntry(
            id = "temp_${Uuid.random()}",
            userId = "",
            entryDate = entry.entryDate,
            durationMinutes = entry.durationMinutes,
            quality = entry.quality,
            bedtime = entry.bedtime,
            wakeTime = entry.wakeTime,
            wakeUps = entry.wakeUps,
            sleepLatencyMinutes = null,
            deepSleepMinutes = null,
            lightSleepMinutes = null,
            remSleepMinutes = null,
            source = null,
            notes = entry.notes,
            loggedAt = Clock.System.now().toString(),
            createdAt = Clock.System.now().toString(),
        )
}

File: mobile/iosApp/Bissbilanz/Persistence/SleepRepository.swift
import Foundation
import Observation
import SwiftData

/// Local-first repository for sleep entries. Reads come from SwiftData
/// (sorted by entry date, newest first); `refresh()` upserts the server list
/// by id and drops rows deleted elsewhere. Writes are SwiftData-first with
/// the upload queued via the sync manager.
@MainActor
@Observable
final class SleepRepository {
    private let context: ModelContext
    private let api: BissbilanzAPI
    private let appMode: AppModeManager
    private let syncManager: SyncManager

    init(context: ModelContext, api: BissbilanzAPI, appMode: AppModeManager, syncManager: SyncManager) {
        self.context = context
        self.api = api
        self.appMode = appMode
        self.syncManager = syncManager
    }

    // MARK: - Reads (local)

    func entries() -> [SleepEntry] {
        let descriptor = FetchDescriptor<LocalSleepEntry>(sortBy: [
            SortDescriptor(\.entryDate, order: .reverse),
        ])
        let rows = (try? context.fetch(descriptor)) ?? []
        return rows.compactMap { $0.toSleepEntry() }
    }

    /// One page of entries, newest first — backs the paginated history list.
    func entries(offset: Int, limit: Int) -> [SleepEntry] {
        var descriptor = FetchDescriptor<LocalSleepEntry>(sortBy: [
            SortDescriptor(\.entryDate, order: .reverse),
        ])
        descriptor.fetchOffset = offset
        descriptor.fetchLimit = limit
        let rows = (try? context.fetch(descriptor)) ?? []
        return rows.compactMap { $0.toSleepEntry() }
    }

    func latest() -> SleepEntry? {
        var descriptor = FetchDescriptor<LocalSleepEntry>(sortBy: [
            SortDescriptor(\.entryDate, order: .reverse),
        ])
        descriptor.fetchLimit = 1
        return (try? context.fetch(descriptor))?.first?.toSleepEntry()
    }

    /// The entry whose day is nearest to `date` ("yyyy-MM-dd") in either
    /// direction, preferring the on-or-before side on a tie.
    func closest(to date: String) -> SleepEntry? {
        var beforeDescriptor = FetchDescriptor<LocalSleepEntry>(
            predicate: #Predicate { $0.entryDate <= date },
            sortBy: [SortDescriptor(\.entryDate, order: .reverse)]
        )
        beforeDescriptor.fetchLimit = 1
        var afterDescriptor = FetchDescriptor<LocalSleepEntry>(
            predicate: #Predicate { $0.entryDate > date },
            sortBy: [SortDescriptor(\.entryDate, order: .forward)]
        )
        afterDescriptor.fetchLimit = 1
        let before = (try? context.fetch(beforeDescriptor))?.first?.toSleepEntry()
        let after = (try? context.fetch(afterDescriptor))?.first?.toSleepEntry()
        guard let before else { return after }
        guard let after else { return before }
        return DateFormatting.dayDistance(before.entryDate, date)
            <= DateFormatting.dayDistance(after.entryDate, date) ? before : after
    }

    // MARK: - Refresh (API → store)

    func refresh() async throws {
        guard !appMode.isLocal else { return }
        let fetched = try await api.getSleepEntries()
        let serverIds = Set(fetched.map(\.id))
        // Rows with an un-uploaded queued write must survive the server
        // response: a refresh racing the sync-queue upload would otherwise
        // reapply the stale server copy over the user's edit (see
        // EntryRepository.refresh, PR #416).
        let pendingIds = syncManager.pendingAffectedIds(table: "sleep")
        for stale in entries() where !serverIds.contains(stale.id)
            && !LocalStore.isTempId(stale.id) && !pendingIds.contains(stale.id)
        {
            deleteRow(id: stale.id)
        }
        for entry in fetched where !pendingIds.contains(entry.id) {
            upsert(entry)
        }
        save()
    }

    // MARK: - Writes (local first + queued upload)

    @discardableResult
    func createEntry(_ create: SleepCreate) async throws -> SleepEntry {
        let temp = makeEntry(from: create, id: LocalStore.makeTempId())
        upsert(temp)
        save()
        syncManager.enqueue(.createSleep(body: create, localId: temp.id))
        return temp
    }

    /// See `EntryRepository.updateEntry` — a missing local row is reported as a
    /// failure without also queueing an upload the caller was just told failed.
    @discardableResult
    func updateEntry(id: String, _ update: SleepUpdate) async throws -> SleepEntry {
        guard let row = fetchRow(id: id), let existing = row.toSleepEntry() else {
            throw APIError.notFound
        }
        let patch = (try? JSONPatch.dictionary(of: update)) ?? [:]
        let updated = (try? JSONPatch.merged(SleepEntry.self, base: existing, patch: patch)) ?? existing
        row.update(from: updated)
        save()
        if LocalStore.isTempId(id) {
            coalesceQueuedCreate(tempId: id, update: update)
        } else {
            syncManager.enqueue(.updateSleep(id: id, body: update))
        }
        return updated
    }

    func deleteEntry(id: String) async throws {
        deleteRow(id: id)
        save()
        if LocalStore.isTempId(id) {
            syncManager.removeQueued(table: "sleep", affectedId: id)
        } else {
            syncManager.enqueue(.deleteSleep(id: id))
        }
    }

    /// Rewrites the still-queued create for a temp-id sleep entry so the
    /// eventual upload carries the edited values.
    private func coalesceQueuedCreate(tempId: String, update: SleepUpdate) {
        for row in syncManager.queuedOperations(table: "sleep", affectedId: tempId) {
            guard let operation = row.operation(),
                  case let .createSleep(body, localId) = operation
            else { continue }
            let patch = (try? JSONPatch.dictionary(of: update)) ?? [:]
            let merged = (try? JSONPatch.merged(SleepCreate.self, base: body, patch: patch)) ?? body
            syncManager.replace(row, with: .createSleep(body: merged, localId: localId))
        }
    }

    // MARK: - Conversion helpers

    private func makeEntry(from create: SleepCreate, id: String) -> SleepEntry {
        let now = ISO8601DateFormatter().string(from: Date())
        return SleepEntry(
            id: id,
            userId: "",
            entryDate: create.entryDate,
            durationMinutes: create.durationMinutes,
            quality: create.quality,
            bedtime: create.bedtime,
            wakeTime: create.wakeTime,
            wakeUps: create.wakeUps,
            sleepLatencyMinutes: nil,
            deepSleepMinutes: nil,
            lightSleepMinutes: nil,
            remSleepMinutes: nil,
            source: nil,
            notes: create.notes,
            loggedAt: now,
            createdAt: now,
            updatedAt: nil
        )
    }

    // MARK: - Store helpers

    private func fetchRow(id: String) -> LocalSleepEntry? {
        var descriptor = FetchDescriptor<LocalSleepEntry>(predicate: #Predicate { $0.id == id })
        descriptor.fetchLimit = 1
        return (try? context.fetch(descriptor))?.first
    }

    private func upsert(_ entry: SleepEntry) {
        if let row = fetchRow(id: entry.id) {
            row.update(from: entry)
        } else {
            context.insert(LocalSleepEntry(entry: entry))
        }
    }

    private func deleteRow(id: String) {
        if let row = fetchRow(id: id) {
            context.delete(row)
        }
    }

    private func save() {
        try? context.save()
        WidgetSnapshotWriter.scheduleUpdate(context: context)
        syncLatestToHealth()
    }

    // MARK: - Apple Health write-back

    /// Dates `HealthKitImporter` is currently writing nights for, read *from*
    /// Health, with the number of import scopes currently covering each. See the
    /// weight repository's twin.
    @ObservationIgnored private var healthImportDateCounts: [String: Int] = [:]

    /// Runs `body` with the Health write-back suppressed for `dates` — imported
    /// nights are marked as already synced rather than written back as
    /// duplicates of the Watch's own sleep samples.
    ///
    /// Reference-counted, because overlapping imports are the normal case:
    /// `BissbilanzApp` runs `importAllIfEnabled` on every foreground
    /// activation while the Sleep tab independently imports from its own `.task`
    /// and pull-to-refresh. With a plain set cleared in `defer`, the inner
    /// scope's exit re-enabled write-back while the outer one was still
    /// writing — producing exactly the app-authored duplicates of the
    /// device's own samples that this suppression exists to prevent.
    func withHealthImportInProgress<T>(dates: Set<String>, _ body: @MainActor () async throws -> T) async rethrows -> T {
        for date in dates {
            healthImportDateCounts[date, default: 0] += 1
        }
        defer {
            for date in dates {
                guard let count = healthImportDateCounts[date] else { continue }
                if count <= 1 {
                    healthImportDateCounts.removeValue(forKey: date)
                } else {
                    healthImportDateCounts[date] = count - 1
                }
            }
        }
        return try await body()
    }

    /// Pushes the newest night to Apple Health on every store mutation. This used
    /// to live in the sleep edit sheet, so a night logged on the Apple Watch or
    /// arriving through `refresh()` never reached Health.
    ///
    /// Nights without both a bedtime and a wake time are skipped — Health needs a
    /// real interval and a fabricated one would pollute it.
    private func syncLatestToHealth() {
        guard let latest = latest(),
              let bedtimeString = latest.bedtime,
              let wakeTimeString = latest.wakeTime,
              let bedtime = DateFormatting.isoDateTime(from: bedtimeString),
              let wakeTime = DateFormatting.isoDateTime(from: wakeTimeString)
        else { return }
        let alreadyInHealth = healthImportDateCounts[latest.entryDate] != nil
        Task {
            await HealthKitService.shared.syncLatestSleep(
                bedtime: bedtime,
                wakeTime: wakeTime,
                entryDate: latest.entryDate,
                alreadyInHealth: alreadyInHealth
            )
        }
    }
}
